examples/flow.cpp

```diff
@@ -171,52 +145,54 @@ int main(int argc, char** argv)
         Opm::Runspec runspec( *deck );
         const auto& phases = runspec.phases();
 
-        // Twophase case
+        // Twophase cases
         if( phases.size() == 2 ) {
             // oil-gas
             if (phases.active( Opm::Phase::GAS ))
             {
-                Opm::FlowMainEbos<TTAG(EclFlowGasOilProblem)> mainfunc;
-                return mainfunc.execute(argc, argv, deck, eclipseState );
+                Opm::flowEbosGasOilSetDeck(*deck, *eclipseState);
```

Comment:
I really don't like this use of static variables, and much prefer the SimulatorParameter solution that was used previously. To me this is the main thing holding me back from merging this!

I could be all alone here, if everyone else (in particular @dr-robertk who did the previous refactoring) disagree I'll yield. In any case, the splitting of compile units can be done separately from the removal of SimulatorParameter (and that would be a PR I'd merge right away). Also I don't see why the call to ...SetDeck() must be here, it would be better to pass it to the main function and call it there (or not, which is what I prefer), in any case such a change would make it possible to get the split compile units in now, and discuss the SimulatorParameter removal later.

Reply:
> I really don't like this use of static variables, and much prefer the SimulatorParameter solution that was used previously. To me this is the main thing holding me back from merging this!

global variables in some form are used quite in flow / eWoms and I actually much prefer them to hell-bending generic code to being dogmatic about not using them. in this case this means adding an otherwise useless additional parameter to the simulator class and also hacking this into the eWoms parameter system where it IMO just feels like a "fifth wheel".

Reply:
I know we disagree on this, which is why I asked for third-parties to weigh in...

Reply:
> to weigh in...

Well - I guess I play in the feather-weight class here, but my take on it is that use of global variables makes testing more difficult.

Reply:
be aware that this "feather light class" needs to be dragged along by quite a bit of infrastructure, and that the "normal" eWoms simulations don't use it, i.e. the additional code paths are basically untested anyway.

Further, notice that this functionality is fully optional: if you don't specify external deck objects before instantiating the simulator the deck files will be parsed a second time in the worst case.

Reply:
> be aware that this "feather light class" [...]

I just meant that my *weight* when *weighing in* was quite limited.

Reply:
I should also add that the eWoms parameter and property systems are intended to be generic, i.e. they are supposed to be useful even for code which is not a simulator.

Reply:
ok, s/light/weight/ in my last comment if you want.

Reply:
my .02c; changes including globals are okay one way - removal of such. for the same reasons as @jokim-hove points to, and general code flow sanity. globals are in my view always a lazy hack. it *may* be okay for simple ints and stuff but for complex classes it inevitably leads to deinit order issues at some point for anything managing resources.

Reply:
did you actually have a look at the current solution? Besides it having to exhibit a copy constructor, the type of the `SimulatorParameter` class is completely opaque to the generic simulator code, yet that code needs to jump through quite a few hoops to make it work. (Side note: Classes also are global entities, so there is not much change from this POV.) Besides the intransparency issue, there are real technical problems with the current approach as well: e.g., how can a situation where e.g. two models each want their specific kind of API for `SimulatorParameter` and a problem which  is supposed to be used with different models to be handled nicely? (note that this is not just a thought experiment: e.g. the co2injection problem can be used with the immiscible, pvs, ncp and flash models, the reservoir problem works in conjunction with blackoil, pvs and ncp.)

IMO global variables are only problematic if they are changed frequently during the runtime of the program. Their current set-once-and-forget-about-them kind of usage in opm-material and -- to a lesser extend -- eWoms is okay because in this case they act like "runtime constexpr" variables  after initialization. (I suppose nobody opposes `constexpr` variables.)

Reply:
no, it was entirely a knee-jerk.